### DISCUSSION TITLE
Help Needed: Forge Tests Failing for Smart Contract Withdrawal Functions

### DISCUSSION THREAD

**Discussion by Giulianomlodi** [Cyfrin/foundry-full-course-cu#1135] [Q&A] [answered]
Hello everyone,

I'm currently working on a smart contract project and facing an issue with my tests using Forge. I've written two functions for withdrawing funds from the contract and corresponding tests to ensure they work correctly. However, the tests are failing, and I'm not sure why. I would appreciate any insights or suggestions from the community.

**Smart Contract Functions:**

I have two functions in my contract for withdrawing funds:

withdraw(): Allows the owner to withdraw all funds from the contract.
withdrawTo(address _receiver, uint256 _amount): Allows the owner to withdraw a specified amount to a given address.
Here's the code for these functions:

```
// Withdraw Contract functions
function withdraw() public onlyOwner {
    if (address(this).balance == 0) {
        revert ContractBalanceIsZero();
    }
    uint256 balance = address(this).balance;
    payable(msg.sender).transfer(balance);
}

function withdrawTo(address _receiver, uint256 _amount) public onlyOwner {
    uint256 balance = address(this).balance;

    if (balance < _amount) {
        revert InsufficientBalance();
    }
    payable(_receiver).transfer(_amount);
}

```

**Tests That Are Failing:**

I've written tests to check the failure scenarios for both functions, but they are not passing as expected. The tests are:

testWithdrawalFailureDueToContractIsEmpty: Checks if the withdraw() function reverts when the contract balance is zero.
testWithdrawalToFailureDueToInsufficientBalance: Checks if the withdrawTo() function reverts when the withdrawal amount is more than the contract's balance.
Here's the code for the tests:

```
function testWithdrawalFailureDueToContractIsEmpty() external {
        console.log("CONTRACT BALANCE: %s ", address(lazyPizzeria).balance);
        vm.expectRevert(LazyPizzeria.ContractBalanceIsZero.selector);
        vm.prank(lazyPizzeria.owner());
        lazyPizzeria.withdraw();
    }

    function testWithdrawalToFailureDueToInsufficientBalance() external {
        uint256 depositAmount = 1 ether;
        vm.prank(lazyPizzeria.owner());
        lazyPizzeria.deposit{value: depositAmount}();
        uint256 withdrawalAmount = address(lazyPizzeria).balance + 1 ether;
        console.log(
            "CONTRACT BALANCE AFTER DEPOSIT: %s ",
            address(lazyPizzeria).balance
        );
        console.log("WITHDRAWAL AMOUNT: %s ", withdrawalAmount);
        console.log("MINTER BALANCE PRE WITHDRAWAL: %s ", MINTER.balance);
        vm.expectRevert(LazyPizzeria.InsufficientBalance.selector);
        vm.prank(lazyPizzeria.owner());
        lazyPizzeria.withdrawTo(MINTER, withdrawalAmount);
    }
```
  
    
    **Error Message:**

For both tests, I'm getting the following error message:

`[FAIL. Reason: Call did not revert as expected]
`

**Request for Assistance:**

I'm not sure why these tests are failing. The logic in the functions seems correct, and the tests are designed to trigger the specified reverts. If anyone has experience with similar issues or sees something I might be missing, your input would be greatly appreciated.

Thank you in advance for your help!

**Comment by 0xusmanf** [comment, marked as answer]
Move your `vm.prank` statements in both functions above `vm.expectRevert`, normally the order doesn't matter but when you make call to a function in `vm.prank` then it does matter because then `vm.expectRevert` will react on the call you are making in `vm.prank` which is why your tests are failing.

Or

You can simply get the address of the owner in the setUp() and then use that address to prank instead of getting it from the contract in every prank.

For example:
```solidity
    address owner;

    function setUp() public {
        owner = lazyPizzeria.owner();
    }

    function testWithdrawalFailureDueToContractIsEmpty() external {
        console.log("CONTRACT BALANCE: %s ", address(lazyPizzeria).balance);
        vm.expectRevert(LazyPizzeria.ContractBalanceIsZero.selector);
        vm.prank(owner);
        lazyPizzeria.withdraw();
    }

    function testWithdrawalToFailureDueToInsufficientBalance() external {
        uint256 depositAmount = 1 ether;
        vm.prank(owner);
        lazyPizzeria.deposit{value: depositAmount}();
        uint256 withdrawalAmount = address(lazyPizzeria).balance + 1 ether;
        console.log(
            "CONTRACT BALANCE AFTER DEPOSIT: %s ",
            address(lazyPizzeria).balance
        );
        console.log("WITHDRAWAL AMOUNT: %s ", withdrawalAmount);
        console.log("MINTER BALANCE PRE WITHDRAWAL: %s ", MINTER.balance);
        vm.expectRevert(LazyPizzeria.InsufficientBalance.selector);
        vm.prank(owner);
        lazyPizzeria.withdrawTo(MINTER, withdrawalAmount);
    }
```

**Reply by Giulianomlodi** [reply]
Thanks so much! It worked
Your answer was really helpful to help me improve and understand more.

**Reply by 0xusmanf** [reply]
You are welcome and I am glad that it helped!

**Comment by vigneshbalu96** [comment]
Happy to see the answer ASAP. Thanks @usmanfarooq91 :100:
